core/storage/statistics/gae_models.py

```diff
@@ -587,114 +590,219 @@ def get_versions(cls, exploration_id):
+        else:
+            main_shard_updated = False
+
+        if last_shard_is_main and (main_shard_updated or last_shard_updated):
+            entities_to_put.append(main_shard)
+        else:
+            if main_shard_updated:
+                entities_to_put.append(main_shard)
+            if last_shard_updated:
+                entities_to_put.append(last_shard)
+
+        cls.put_multi(entities_to_put)
 
     @classmethod
-    def _get_entity_id(cls, exploration_id, exploration_version, state_name):
-        return ':'.join([exploration_id,
-                         str(exploration_version),
-                         state_name])
+    def insert_submitted_answers(
+            cls, exploration_id, exploration_version, state_name,
+            interaction_id, new_submitted_answer_dict_list):
+        """Given an exploration ID, version, state name, and interaction ID,
+        attempt to insert a list of specified SubmittedAnswers into this model,
+        performing sharding operations as necessary. This method automatically
+        commits updated/new models to the data store. This method returns
+        nothing. This method can guarantee atomicity since mutations are
+        performed transactionally, but it cannot guarantee uniqueness for answer
+        submission. Answers may be duplicated in cases where a past transaction
+        is interrupted and retried. Furthermore, this method may fail with a
+        DeadlineExceededError if too many answers are attempted for submission
+        simultaneously.
+        """
+        transaction_services.run_in_transaction(
+            cls._insert_submitted_answers_unsafe, exploration_id,
+            exploration_version, state_name, interaction_id,
+            new_submitted_answer_dict_list)
 
-    def save(self):
-        """Commit to storage."""
-        # This may fail if answers_list is too large.
-        try:
-            self.put()
-        except Exception as e:
-            logging.error(e)
+    @classmethod
+    def _get_entity_id(
+            cls, exploration_id, exploration_version, state_name, shard_id):
+        return ':'.join([
+            exploration_id, str(exploration_version), state_name,
+            str(shard_id)])
+
+    @classmethod
+    def _shard_answers(
+            cls, current_answer_list, current_answer_list_size,
+            new_answer_list):
+        """Given a current answer list which can fit within one NDB entity and
+        a list of new answers which need to try and fit in the current answer
+        list, shard the answers such that a list of answer lists are returned.
+        The first entry is guaranteed to contain all answers of the current
+        answer list.
+        """
+        # Sort the new answers to insert in a descending by size in bytes.
+        new_answer_list_sorted = sorted(
+            new_answer_list, key=lambda x: x['json_size'])
+        sharded_answer_lists = [list(current_answer_list)]
```

Comment:
Yes, a copy is needed here because we're appending to the list later and we don't want to append to the list stored within one of the models, since it may not need to change.